This PR implements an adjusted R2 calculation for cca

dmcglinn opened the pull request:
Based on conversations in #161 I decided to provide code for implementing the permutation based approach of adjusting the R2 calculation for cca that was developed by Peres-Neto et al 2006. There are currently some limitations of my implementation (see commit log). I'm happy to work to overcome these limitations ironed out so that this method can be more fully realized, but I'll likely need some suggestions for how to do this.

jarioksa commented:
It seems that I need to do some reading. My impression was that Peres-Neto's method was more complicated, but I never tried to dive through the details of calculation. Your implementation is very direct and simple. Actually, it could be written more briefly using `vegan:::permutest.cca`:
```R
RsquareAdj.cca <-
    function (x, nperm = 1000) 
{
    r2 <- x$CCA$tot.chi/x$tot.chi
    p <- permutest(x, nperm)
    radj <- 1 - ((1-r2) / (1 - mean(p$num/x$tot.chi)))
    list(r.squared = r2, adj.r.squared = radj)
}
```
This is also faster and easier to maintain. However, these approaches should be verified against Peres-Neto's paper. Do you have any test case?

BTW, the main usage of `RsquareAdj` is in embedded usage in functions like `ordiR2step`  and `varpart`. When I enabled your code in `ordiR2step` I got an error:
```R
Error in svd(Xbar, nu = 0, nv = 0) : a dimension is zero
> traceback()
10: stop("a dimension is zero")
9: svd(Xbar, nu = 0, nv = 0)
8: cca.default(d$X, d$Y, d$Z)
7: cca.formula(Y_rand ~ 1, data = dune.env)
6: cca(Y_rand ~ 1, data = dune.env) at <text>#1
5: eval(expr, envir, enclos)
4: eval(parse(text = paste(x$call[1], "(", x$call[2], ", data=", 
       x$call[3], ")", sep = ""))) at RsquareAdj.R#48
3: RsquareAdj.cca(update(object, fla))
2: RsquareAdj(update(object, fla)) at ordiR2step.R#67
1: ordiR2step(m0, scope = formula(m))
```
I think this happens because the function tries to evaluate an empty model with no constrained component  `cca(Y ~ 1)`.

dmcglinn commented:
Hey @jarioksa thanks for the feedback. The `permutest` implementation is a great suggestion. I'll integrate that. Yea the approach is pretty simple. See equ 5 in the Peres-Neto et al 2006 paper. Do you basically envision several `if` statements within this function to catch all the special cases of different ways a cca model could be specified (e.g., the empty model example) or is their a cleaner and more direct way to go?

dmcglinn commented:
Also I don't have a clean test case to compare with the Peres-Neto paper. They did provide some matlab code and an executable. I'll do some manual comparison their to see if the values match.

jarioksa commented:
When I checked the `permutest` implementation, it  passed the `ordiR2step` case: it seems that `permutest.cca` handled smoothly missing `CCA` component. 

We use `RsquareAdj` in `ordiR2step` and in `varpart`, and both currently check the input and refuse to work with `cca` results. Handling of  `cca` results should be accepted, and probably also an example added, and possibly also a test case in `tests/vegan-tests.R` to catch regressions.

dmcglinn commented:
Nice ok well i'll report back on my manuel testing foray

On 04/15/2016 08:23 AM, Jari Oksanen wrote:
>
> When I checked the |permutest| implementation, it passed the 
> |ordiR2step| case: it seems that |permutest.cca| handled smoothly 
> missing |CCA| component.
>
> We use |RsquareAdj| in |ordiR2step| and in |varpart|, and both 
> currently check the input and refuse to work with |cca| results. 
> Handling of |cca| results should be accepted, and probably also an 
> example added, and possibly also a test case in |tests/vegan-tests.R| 
> to catch regressions.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub 
> <https://github.com/vegandevs/vegan/pull/162#issuecomment-210444826>
>

jarioksa commented:
@gavinsimpson : in permutation tests we add observed statistic with permutations, and therefore we have, say, 999 permutations + one observations. What about permutation-based R²? Should we have the observed R² there, too?

dmcglinn commented:
Peres-Neto did not specify one way or another on if the observed R2 should be include. I would lean towards not including because if the number of permutations is small this will bias the result and the permuted R2 represent the explained variance of random variables where as in a permutation test it represents potentially observable shufflings of the data which have slightly different interpretations I think. 

As a test, I used the executable program archived here: http://www.esapubs.org/archive/ecol/E087/158/suppl-3.htm  to compute the cca r2 and adjusted r2 for the `varespec` and `varechem` datasdets. Here is the output from that analysis: 
[cca_r2.txt](https://github.com/vegandevs/vegan/files/221171/cca_r2.txt)

This matches r2 and adjusted r2 computed for my suggested code to the second decimal point with 1000 permutations.

gavinsimpson commented:
I agree with Dan's assessment re including observed R2 in the permutation distribution. Pedro and co are not using the permutations in a hypothesis testing mode (looking for evidence to reject a null) but are simply generating a null distribution from which to compute an estimator for p/(n - 1). Though the distinction is small.

dmcglinn commented:
Hey @jarioksa and @gavinsimpson I think the function `RsquareAdj.cca` is ready to go now. I implemented the version that @jarioksa suggested and I updated the man file accordingly.

gavinsimpson commented on R/RsquareAdj.R:
```diff
@@ -35,11 +35,12 @@
 ## cca result: no RsquareAdj
 RsquareAdj.cca <-
-    function(x, ...)
+    function (x, nperm = 1000) 
```
Function definition needs to have `...` as an argument because the generic has this argument.

For consistency, `nperm` should be renamed to `permutations` as per `anova.cca` etc.

gavinsimpson commented on R/RsquareAdj.R:
```diff
@@ -35,11 +35,12 @@
-    radj <- NA
-    list(r.squared = R2, adj.r.squared = radj)
+    r2 <- x$CCA$tot.chi / x$tot.chi
+    p <- permutest(x, nperm)
```
How about passing `...` to `permutest()` here? That way people can call this with `parallel` or other functions should they need to speed this up.

jarioksa commented:
Yes, we should be able to pass `parallel` to `permutest.cca`: we need it in particular when `RsquareAdj.cca` is embedded in other functions and called repeatedly.

dmcglinn commented:
done

gavinsimpson commented:
Thanks for handling this @dmcglinn. Merged.

dmcglinn commented:
It was great working with both of you thanks for your encouragement and guidance!

gavinsimpson commented:
Cheers @dmcglinn and thanks again for this - @jarioksa and I have spoken about this on and off for a while but we never got round to to doing anything about it. Now I can update my slides ahead of the Vegan workshop at ESA to say this is done!

jarioksa commented:
Thanks @dmcglinn . I have now allowed `cca` in `ordiR2step` in a separate branch (which may need some scrutiny before the release). However, it seems that we still can't implement variation partition (`varpart`) with CCA: we use there a minimal `simpleRDA2` function followed with `RsquareAdj.default`, and the current `RsquareAdj.cca` cannot be pluggled in. (It is surprisingly trivial to estimate unadjusted R squared for CCA using `simpleRDA2, but adjusted R-squared needs more coding.)
